src/main/java/gov/cabinetofice/gapuserservice/service/user/OneLoginUserService.java

```diff
@@ -200,17 +200,19 @@ public void invalidateUserJwt(final Cookie customJWTCookie, final HttpServletRes
         response.addCookie(thirdPartyAuthToken);
     }
 
-    public void validateRoles(List<Role> userRoles) {
+    public void validateRoles(List<Role> userRoles, String payloadRoles) {
         final Set<String> formattedUserRoles = userRoles.stream()
                 .map(role -> roleMapper.roleToRoleDto(role).getName())
                 .collect(Collectors.toSet());
-        if(formattedUserRoles.isEmpty()){
-            throw new UnauthorizedException("User is blocked");
+        boolean userHasBeenUnblocked = payloadRoles.equals("[]") && formattedUserRoles.size() > 0;
+
+        if(formattedUserRoles.isEmpty() || userHasBeenUnblocked){
+            throw new UnauthorizedException("Payload is invalid - User is blocked or has been unblocked");
```

Comment:
Why is an unblocked user seeing an exception?

Reply:
Because their payload is out of sync with the roles in the db, this exception will send them back to /login which refreshes the payload

Reply:
What would be the step to reproduce this?
1. Admin is blocked
2. Logged out
3. Tries to relog before being unblocked
4. Gets unblocked while their session is still active ?

Reply:
yes